### PR TITLE
[WIP] refactor: WorkerBundle

### PR DESCRIPTION
**Pull request by xtuc**
refactor: WorkerBundle

This change introduce the notion of the WorkerBundle. Conceptually it
represents result of the backend (javascript, rust or webpack) computation.

The abstraction provide generalization for creating metadata and
uploading the worker to the API using a multipart form.

To maintain compatiblity with the rust backend, the WorkerBundle can be
contructed with a file in `metadata_path` property. We will use the file
directly instead of generating the metadata from the bindings.

For easier testing WorkerBundle allow to pass a `out_root` property
which will be used as base directory when persisting the bundle files.

Fixes https://github.com/cloudflare/wrangler/issues/293

### CODE REVIEW

File: src/commands/build/wranglerjs/output.rs [hunk 1]
```diff
@@ -77,3 +125,119 @@ mod tests {
         assert_eq!(wranglerjs_output.wasm_size(), "3 bytes");
     }
 }
+
```
Comment (cwndrws) [inline src/commands/build/wranglerjs/output.rs]: Probably shouldn't check in commented out tests

File: src/commands/build/mod.rs [hunk 2]
```diff
@@ -1,17 +1,29 @@
+// FIXME: move
```
Comment (cwndrws) [inline src/commands/build/mod.rs]: It's not clear what needs to happen here.

File: src/worker_bundle.rs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+use crate::settings::metadata;
+use crate::terminal::message;
+
+const WORKER_BUNDLE_OUT: &str = "./worker";
+
+#[derive(Debug)]
+pub struct WorkerBundle {
+    // out: String,
+    pub script_path: PathBuf,
+    pub wasm_path: Option<PathBuf>,
```
Comment (cwndrws) [inline src/worker_bundle.rs]: It seems like the wasm stuff should get owned by a binding rather than being top-level attributes of a WorkerBundle.

File: src/commands/build/wranglerjs/output.rs [hunk 4]
```diff
@@ -20,33 +29,72 @@ pub struct WranglerjsOutput {
...
-    pub fn wasm_size(&self) -> String {
-        let size = self.wasm.to_owned().unwrap().len();
-        match NumberPrefix::decimal(size as f64) {
-            Standalone(bytes) => format!("{} bytes", bytes),
-            Prefixed(prefix, n) => format!("{:.0} {}B", n, prefix),
-        }
-    }
+// We inject some code at the top-level of the Worker; called {prologue}.
+// This aims to provide additional support, for instance providing {window}.
+pub fn create_prologue() -> String {
```
Comment (cwndrws) [inline src/commands/build/wranglerjs/output.rs]: This seems out of scope for this change.

File: src/commands/build/mod.rs (hunk 2 above)
Comment (xtuc) [reply src/commands/build/mod.rs]: We can move the krate and package code into build because it's captured by WorkerBundle.

File: src/commands/build/wranglerjs/output.rs (hunk 4 above)
Comment (xtuc) [reply src/commands/build/wranglerjs/output.rs]: I just moved the code from the bundle.rs file, that I deleted in favor of the shared WorkerBundle.

File: src/worker_bundle.rs (hunk 3 above)
Comment (xtuc) [reply src/worker_bundle.rs]: Your approach make sense to me now. I was hesitant to modify the binding struct because it doesn't strictly match with the online bindings anymore, but I guess it's fine.

**Comment by ashleymichal** [comment]
closing in favor of incoming pr

**Comment by xtuc** [comment]
What was the issue with this PR?

**Comment by ashleymichal** [comment]
Thanks for asking, I should have included a proper explanation:

We stepped back and worked out a design to solve the problem, including identifying a few refactors that will be beneficial. Because of the scope of the design we came up with, and the steps we identified, we felt it would be easier to start off of the master branch, using these two PRs (this and #299 ) as reference.
